optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/reproducer/DroolsReproducer.java

```diff
@@ -0,0 +1,312 @@
+    // Test printing
+    //------------------------------------------------------------------------------------------------------------------
+    //
+    private void printReproducer() {
+        printInit();
+        printSetup();
+        printTest();
+    }
+
+    private void printInit() {
+        String domainPackage = null;
+        for (Fact fact : journal.getFacts()) {
+            Class<? extends Object> factClass = fact.getInstance().getClass();
+            for (Annotation ann : factClass.getAnnotations()) {
+                if (PlanningEntity.class.equals(ann.annotationType())) {
+                    domainPackage = factClass.getPackage().getName();
+                }
+                break;
+            }
+            if (domainPackage != null) {
+                break;
+            }
+        }
+
+        if (domainPackage == null) {
+            throw new IllegalStateException("Cannot determine planning domain package.");
+        }
+
+        reproducerLog.info("package {};\n", domainPackage);
+        SortedSet<String> imports = new TreeSet<String>();
+        imports.add("org.junit.Before");
+        imports.add("org.junit.Test");
+        imports.add("org.kie.api.KieServices");
+        imports.add("org.kie.api.builder.KieFileSystem");
+        imports.add("org.kie.api.builder.model.KieModuleModel");
+        imports.add("org.kie.api.io.ResourceType");
+        imports.add("org.kie.api.runtime.KieContainer");
+        imports.add("org.kie.api.runtime.KieSession");
+        for (Fact fact : journal.getFacts()) {
+            for (Class<?> cls : fact.getImports()) {
+                if (!cls.getPackage().getName().equals(domainPackage)) {
+                    imports.add(cls.getCanonicalName());
+                }
+            }
+        }
+
+        for (String cls : imports) {
+            reproducerLog.info("import {};", cls);
+        }
+        reproducerLog.info(
+                "\n" +
+                "public class DroolsReproducerTest {\n" +
+                "\n" +
+                "    KieSession kieSession;");
+        for (Fact fact : journal.getFacts()) {
+            fact.printInitialization(reproducerLog);
+        }
+        reproducerLog.info("");
+    }
+
+    private void printSetup() {
+        reproducerLog.info(
+                "    @Before\n" +
+                "    public void setUp() {\n" +
+                "        KieServices kieServices = KieServices.Factory.get();\n" +
+                "        KieModuleModel kieModuleModel = kieServices.newKieModuleModel();\n" +
+                "        KieFileSystem kfs = kieServices.newKieFileSystem();\n" +
+                "        kfs.writeKModuleXML(kieModuleModel.toXML());\n" +
+                // TODO don't hard-code score DRL
+                "        kfs.write(kieServices.getResources().newClassPathResource(\"org/optaplanner/examples/nurserostering/solver/nurseRosteringScoreRules.drl\").setResourceType(ResourceType.DRL));\n" +
```

Comment:
This needs to be resolved. The ideal solution would be to get all rules (not the source files) through KIE API, prune them automatically and finally inline source of the minimal set of rules that still reproduces the bug. I don't know if that's possible.